### PR TITLE
Add Inner Trait pattern

### PR DESCRIPTION
**Pull request by rbran**

### CODE REVIEW

**Comment by burdges** [comment]
You'd avoid re-implementing the exposed methods if you require `InnerCar : Car` btw.

It's a useful pattern if you've very large crates or internally use `dyn InnerCar` lots.  It'll also become more useful whenever specialization arrives.

Yet, if your project has many small crates, then you'd typically achieve this using free `fn`s instead.
```
fn set_speed<C: Car>(car: &mut C, new_speed: f64) { ... }
fn get_acceleration<C: Car>(car: &mut C) -> f64 { ... }
```

File: patterns/behavioural/inner-trait.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+//more Car implementations...
+```
+
+## Motivation
+
+This pattern allows developers to provide additional functionality to the
+implementation of a public trait without exposing that functionality as part of
+the public API. By using a private trait, developers can also improve the
+reusability of their code, since the private functionality can be reused across
+multiple implementations of the public trait.
```
Comment (neithernut) [inline patterns/behavioural/inner-trait.md:74]: Suppose I'm using a trait `Car` from another crate `vehicles`, but for a very specific purpose requiring some additional code. Do you feel that a new trait `InnerCar` private to my own crate would fall under this pattern? If yes, this would be an alternative motivation, wouldn't it?

**Review by neithernut — COMMENTED** [review]
I feel this is a specific manifestation of a more general pattern which may or may not warrant its own (sub)section including #310 and "type classes" (note the double quotes) as in

```rust
trait Foo {
    // ...
}

impl <T: ...> Foo for T {
    // ...
}
```

So maybe we want to broaden this into the "extension trait" pattern proposed in #310. If not we should draw a line bwtween those two at some point in the future (as in, if/when someone writes a section about "extension traits" or something similar).

TL;DR: LGTM, no objections against merging this as is.

File: patterns/behavioural/inner-trait.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+In rust there is the public/private trait analogous, the private trait
+implementing the hidden functionalities, while the public trait exposes
+functionality.
+
+## See also
+
+Wikipedia [OOP Interface](https://en.wikipedia.org/wiki/Interface_%28object-oriented_programming%29).
+
+Blog post from [Predrag](https://web.archive.org/web/20230406211349/https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/)
+about sealed, private and other patterns for traits.
```
Comment (pickfire) [inline patterns/behavioural/inner-trait.md:105]: I think once the other PR is merged, we need to link them to each others.

File: patterns/behavioural/inner-trait.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+It is possible to define a private trait that implements all the
+methods of a public trait and also includes some private functions. This pattern
+can be used to provide additional functionality to the implementation of a
+public trait while keeping the private methods hidden from the public API.
+
+## Example
+
+This example demonstrate how a public trait `Car` can be implemented and include
+extra private methods using a auxiliary private trait `InnerCar`.
+
```
Comment (pickfire) [inline patterns/behavioural/inner-trait.md:14]: I think we probably want to show a real example in standard library or in popular crates for this.

File: patterns/behavioural/inner-trait.md (hunk 1 above)
Comment (pickfire) [reply patterns/behavioural/inner-trait.md:74]: There is another layer on top of this is to prevent breaking changes, since we can now limit functionality without exposing public API.

File: patterns/behavioural/inner-trait.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+```rust,ignore
+// trait that is public and part of the API
+pub trait Car {
+    fn get_speed(&self) -> f64;
```
Comment (pickfire) [inline patterns/behavioural/inner-trait.md:18]: I wouldn't recommend using `get_` since `speed` will do, this looks javaish and make it unnecessary long for not much benefits, rust already have types for this, we don't have to spell it out in function name.

File: patterns/behavioural/inner-trait.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    fn accelerate(&mut self, duration: f64);
+    fn brake(&mut self, force: f64);
```
Comment (pickfire) [inline patterns/behavioural/inner-trait.md:20]: I would recommend keeping this to at most 2 functions to make a minimal example.

**Comment by rbran** [comment]
The more that I think about this pattern, the less I like it. It fells too complex/edge-case for a pattern, or maybe is a simple adaptation/equivalent for an "abstract-class" made by someone that is too stuck with the OOP mentality.

**Comment by rbran** [comment]
I'll rethink this, if I got a more concise idea for a pattern, I'll create a new PR.
